### PR TITLE
James 2362

### PR DESCRIPTION
**Pull request by rouazana**
WIP about flags (answered, forwarded) update handling

### CODE REVIEW

File: server/protocols/jmap-integration-testing/memory-jmap-integration-testing/src/test/java/org/apache/james/jmap/memory/MemorySetMessagesMethodTest.java [hunk 1]
```diff
@@ -48,4 +50,11 @@ protected void await() {
     protected MessageId randomMessageId() {
         return new InMemoryMessageId.Factory().fromString(String.valueOf(random.nextInt(100000) + 100));
     }
+
+    @Ignore("JAMES-2363 Bug: memory implementation should preserver mime message ids")
+    @Override
+    @Test
+    public void mimeMessageIdShouldBePreservedWhenSending() {
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/memory-jmap-integration-testing/src/test/java/org/apache/james/jmap/memory/MemorySetMessagesMethodTest.java]: TODO

Comment (chibenwa) [reply server/protocols/jmap-integration-testing/memory-jmap-integration-testing/src/test/java/org/apache/james/jmap/memory/MemorySetMessagesMethodTest.java]: I found the bug after a debbugging session, and **it's tricky!**

The problem is that **RemoveHeader** mailet calls **MimeMessage::saveChanges()**, which in turn calls **updateHeaders**, which then generate a new **Message-Id**....

Comment (chibenwa) [reply server/protocols/jmap-integration-testing/memory-jmap-integration-testing/src/test/java/org/apache/james/jmap/memory/MemorySetMessagesMethodTest.java]: An alternative approach (less invasive, but relies on inheritance) would be to rely on a custom implementation of **updateHeaders()** in **MimeMessageCopyOnWriteProxy**

Comment (chibenwa) [reply server/protocols/jmap-integration-testing/memory-jmap-integration-testing/src/test/java/org/apache/james/jmap/memory/MemorySetMessagesMethodTest.java]: It is really depressing, these methods with unexpected side effects are called everywhere... Even a **setHeader** on MimeMessageCopyOnWriteProxy triggers this...

Comment (chibenwa) [reply server/protocols/jmap-integration-testing/memory-jmap-integration-testing/src/test/java/org/apache/james/jmap/memory/MemorySetMessagesMethodTest.java]: I finally ended up replacing **MimeMessage::saveChanges()** calls by a **MimeMessageSaver::save()** which preserves message id.

**Comment by chibenwa** [comment]
```
[bec4c1618d5cd88669a33c9c7ec6a61cd57cdbb1] Tests in error: 
[bec4c1618d5cd88669a33c9c7ec6a61cd57cdbb1]   UseHeaderRecipientsTest.serviceShouldGhostEmail:113 » Messaging IOException wh...
[bec4c1618d5cd88669a33c9c7ec6a61cd57cdbb1]   UseHeaderRecipientsTest.serviceShouldResendTheEmail:129 » Messaging IOExceptio...
[bec4c1618d5cd88669a33c9c7ec6a61cd57cdbb1]   UseHeaderRecipientsTest.serviceShouldSetEmptyRecipientWhenNoRecipientsInTheMimeMessage:100 » Messaging
[bec4c1618d5cd88669a33c9c7ec6a61cd57cdbb1]   UseHeaderRecipientsTest.serviceShouldSetMimeMessageRecipients:70 » Messaging I...
[bec4c1618d5cd88669a33c9c7ec6a61cd57cdbb1]   UseHeaderRecipientsTest.serviceShouldSetToCcAndBccSpecifiedInTheMimeMessage:86 » Messaging
[bec4c1618d5cd88669a33c9c7ec6a61cd57cdbb1]   UseHeaderRecipientsTest.serviceShouldSupportAddressList:159 » Messaging IOExce...
[bec4c1618d5cd88669a33c9c7ec6a61cd57cdbb1]   UseHeaderRecipientsTest.serviceShouldSupportMailboxes:173 » Messaging IOExcept...
```

File: javax-mail-extension/src/main/java/org/apache/james/javax/MimeMessageBuilder.java [hunk 2]
```diff
@@ -380,7 +380,6 @@ public MimeMessage build() throws MessagingException {
         for (Header header: headerList) {
             mimeMessage.addHeader(header.name, header.value);
         }
-        mimeMessage.saveChanges();
```
Comment (rouazana) [inline javax-mail-extension/src/main/java/org/apache/james/javax/MimeMessageBuilder.java]: unrelated to this commit

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java [hunk 3]
```diff
@@ -82,10 +82,6 @@ public void replaceMailContentAndSubject(Mail mail, ReplaceConfig replaceConfig,
         try {
             boolean subjectChanged = applySubjectReplacingUnits(mail, replaceConfig, charset);
             boolean contentChanged = applyBodyReplacingUnits(mail, replaceConfig, charset);
-
-            if (subjectChanged || contentChanged) {
-                mail.getMessage().saveChanges();
-            }
```
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: unrelated to this commit

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/MailModifier.java [hunk 4]
```diff
@@ -210,12 +210,13 @@ public void initializeDateIfNotPresent() throws MessagingException {
...
 
-    public void setMessageId(Mail originalMail) throws MessagingException {
-        String messageId = originalMail.getMessage().getMessageID();
-        if (messageId != null) {
-            mail.getMessage().setHeader(RFC2822Headers.MESSAGE_ID, messageId);
+    public void setMessageId(String savedValue) throws MessagingException {
+        if (savedValue != null) {
+            mail.getMessage().setHeader(RFC2822Headers.MESSAGE_ID, savedValue);
+            String messageID = mail.getMessage().getMessageID();
+            System.out.println(messageID);
```
Comment (rouazana) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/MailModifier.java]: debug

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
I took time for adding more unit tests and also better cleaning the history.

File: javax-mail-extension/src/main/java/org/apache/james/javax/MimeMessageSaver.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.util.Optional;
+
+import javax.mail.MessagingException;
+import javax.mail.internet.MimeMessage;
+
+import com.github.fge.lambdas.Throwing;
+
+public class MimeMessageSaver {
+    public static void save(MimeMessage mimeMessage) throws MessagingException {
```
Comment (mbaechler) [inline javax-mail-extension/src/main/java/org/apache/james/javax/MimeMessageSaver.java]: what does it mean to "save" a MimeMessage ? the method has a input but no output, so could we describe what is the expected side-effect ?

File: javax-mail-extension/src/main/java/org/apache/james/javax/MimeMessageSaver.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        String messageID = mimeMessage.getMessageID();
+        mimeMessage.saveChanges();
+        Optional.ofNullable(messageID)
+            .ifPresent(Throwing.consumer((String id) -> mimeMessage.setHeader("Message-ID", id)).sneakyThrow());
```
Comment (mbaechler) [inline javax-mail-extension/src/main/java/org/apache/james/javax/MimeMessageSaver.java]: in this case, the message has been half-changed but the client can continue to use it, right ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/MailModifier.java [hunk 7]
```diff
@@ -210,12 +210,11 @@ public void initializeDateIfNotPresent() throws MessagingException {
         }
     }
 
-    public void setMessageId(Mail originalMail) throws MessagingException {
-        String messageId = originalMail.getMessage().getMessageID();
-        if (messageId != null) {
-            mail.getMessage().setHeader(RFC2822Headers.MESSAGE_ID, messageId);
+    public void setMessageId(String savedValue) throws MessagingException {
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/MailModifier.java]: what means savedValue in this context ?

File: mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java [hunk 8]
```diff
@@ -107,6 +108,14 @@ public FakeMailContext build() {
 
     public static class SentMail {
 
+        private static MimeMessage tryCopyMimeMessage(MimeMessage msg) throws MessagingException {
+            try {
+                return new MimeMessage(msg);
+            } catch (Exception e) {
+                return MimeMessageBuilder.mimeMessageBuilder().build();
```
Comment (mbaechler) [inline mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java]: I can't understand this part.

Comment (chibenwa) [reply mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java]: A null message generated an exception upon the copy. I provided a default value. But

```
Optional.ofNullable(msg)
    .map(MimeMessage::new)
    .orElse(null);
```

Indeed seems better. We should rather use it.

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/MailModifier.java (hunk 7 above)
Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/MailModifier.java]: s/savedValue/savedMessageIdValue/

**Comment by chibenwa** [comment]
I had a look at **MimeMessageWrapper**.

It seems that it overrides **MimeMessage** in order to keep a similar behaviour, but avoid overriding it's **MessageID**.

Maybe we could enforce always dealing with a MimeMessageWrapper? (Enforce it as part of FakeMail and MailImpl implementations?). 

Maybe this approach would be cleaner....

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java [hunk 9]
```diff
@@ -64,6 +64,11 @@
         From, To, Cc, Bcc
     }
 
+    public enum Precision {
+        Exact,
+        Imprecise
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java]: I prefer `Approximate`

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java [hunk 10]
```diff
@@ -765,6 +770,8 @@ public static Criterion all() {
 
     private List<Sort> sorts = Collections.singletonList(new Sort(Sort.SortClause.Uid, Sort.Order.NATURAL));
 
+    private Precision exact;
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java]: What is the default precision?

**Comment by chibenwa** [comment]
We will need a manual PUT_MAPPING on prod to add the new "MimeMessageID" field (otherwise it will be indexed with **standard** analyser and all this work would have been worthless.

See https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-mapping.html

```
PUT mailbox/_mapping/message
{
  "properties": {
    "name": {
      "properties": {
        "mimeMessageID": { 
          "type": "string",
          "index": "not_analyzed"
        }
      }
    }
  }
}
```

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**